tempodb/encoding/vparquet2/block_traceql.go

```diff
@@ -1874,3 +1868,25 @@ func newSpanAttr(name string) traceql.Attribute {
 func newResAttr(name string) traceql.Attribute {
 	return traceql.NewScopedAttribute(traceql.AttributeScopeResource, false, name)
 }
+
+func unionIfNeeded(definitionLevel int, iters []parquetquery.Iterator, pred parquetquery.GroupPredicate) parquetquery.Iterator {
+	switch len(iters) {
+	case 0:
+		return nil
+	case 1:
+		return iters[0]
+	default:
+		return parquetquery.NewUnionIterator(definitionLevel, iters, pred)
+	}
+}
+
+func orIfNeeded(preds []parquetquery.Predicate) parquetquery.Predicate {
```

Comment:
i was noticing that `{ span.foo = "bar" && span.foo = "baz" }` creates the following:

```
LeftJoinIterator: 4: attributeCollector{}
required: 
	SyncIterator: rs.list.element.ss.list.element.Spans.list.element.Attrs.list.element.Key : InstrumentedPredicate{8, StringInPredicate{foo, foo, }}
optional: 
	SyncIterator: rs.list.element.ss.list.element.Spans.list.element.Attrs.list.element.Value : InstrumentedPredicate{0, OrPredicate{StringInPredicate{bar, },StringInPredicate{bat, }}}
```

- for the key iterator it's kind of an edge case, but we could write guard code that prevented adding dupes
- for the val iterator i'm wondering if a StringInPredicate{bar, bat, } works better? this behavior is the same as in main, so might not be worth addressing. unsure if it's even possible.

Reply:
Yep this idea came up in the original vParquet and most of our predicates are designed for batches (RegexIn, StringIn).  They should be faster than the Or, but this type of optimization just hasn't made it to the top of the list.